### PR TITLE
Change CPU clustering to use flat (alt) edm

### PR DESCRIPTION
**Pull request by guilhermeAlmeida1**
This PR almost fully eliminates the use of all the "outdated" jagged EDM in the clusterisation.
It is quite a lengthy PR as it really touches almost everything in our code, but there are still a few bits and pieces in our repository using the old edm, which is why I didn't just delete all the old headers just yet.

I'll admit that I made the physics performance "mapper" functions a bit uglier, but this whole piece of code was already a mess and the solution I implemented was the only one I could see without fully rewriting this from scratch.

### CODE REVIEW

**Comment by beomki-yeo** [comment]
> I'll admit that I made the physics performance "mapper" functions a bit uglier, but this whole piece of code was already a mess and the solution I implemented was the only one I could see without fully rewriting this from scratch.

You can mess it up as much as you want. I will rewrite them in the near future

**Comment by guilhermeAlmeida1** [comment]
Performance-wise, the cpu-only throughput doens't change a lot. At low pile-up it actually becomes a bit faster with the flat edm but at high pile-up it's practically the same.

File: core/include/traccc/edm/internal_spacepoint.hpp [hunk 1]
```diff
@@ -23,16 +23,7 @@ namespace traccc {
...
-    using spacepoint_container_type = host_container<geometry_id, spacepoint_t>;
-    using link_type = typename spacepoint_container_type::link_type;
-
-    link_type m_link;
-
-    /// TODO: This is a *VERY* temporary variable which will be removed asap.
-    /// only here while transitioning from the jagged to non-jagged
-    /// clusterization and prevent even more code duplication
-    unsigned int m_link_alt;
+    unsigned int m_link;
```
Comment (krasznaa) [inline core/include/traccc/edm/internal_spacepoint.hpp:26]: I would prefer to keep the

```c++
using link_type = unsigned int;
```

declaration, and then using `link_type` in the code, sort of as before. Would reduce the amount of necessary changes a little.

File: core/src/clusterization/measurement_creation.cpp [hunk 2]
```diff
@@ -36,25 +36,24 @@ measurement_creation::output_type measurement_creation::operator()(
...
-            result.push_back(cell_module(module),
-                             measurement_collection_types::host(&(m_mr.get())));
-        }
-
         // Get the cluster.
         cluster_container_types::host::item_vector::const_reference cluster =
             clusters.at(i).items;
 
+        if (cluster.size() == 0) {
+            // this should never happen, but included for safekeeping.
```
Comment (krasznaa) [inline core/src/clusterization/measurement_creation.cpp:44]: :thinking: If it should really never happen, then let's put an assertion here. I.e.

```c++
assert(cluster.size() > 0);
```

File: device/cuda/src/cca/component_connection.cu [hunk 3]
```diff
@@ -37,7 +37,7 @@ struct cell_container {
     channel_id* channel1 = nullptr;
     scalar* activation = nullptr;
     scalar* time = nullptr;
-    geometry_id* module_id = nullptr;
+    unsigned int* module_id = nullptr;
```
Comment (krasznaa) [inline device/cuda/src/cca/component_connection.cu:40]: Did I miss something? Why did you remove the `geometry_id` type? :confused: Typedefs are generally good in this type of code.

File: examples/options/include/traccc/options/particle_gen_options.hpp [hunk 4]
```diff
@@ -63,7 +63,7 @@ struct particle_gen_options {
         mom_range = vm["gen-mom-gev"].as<Reals<scalar_t, 2>>();
         phi_range = vm["gen-phi-degree"].as<Reals<scalar_t, 2>>();
         const auto eta_range = vm["gen-eta"].as<Reals<scalar_t, 2>>();
-        // TODO: remove the conversion here...
+        //  remove the conversion here...
```
Comment (krasznaa) [inline examples/options/include/traccc/options/particle_gen_options.hpp:66]: I believe this change went into the PR by mistake...

**Review by krasznaa — CHANGES_REQUESTED** [review]
I largely like it. Though admittedly I was really only looking for bad code smells in general. The PR is way too big to properly check it line-by-line... (At the same time, I didn't see an obvious way either of how to break it into multiple pieces.)

Hopefully we're now getting close to removing/renaming all the `alt_` prefixed types...

File: device/cuda/src/cca/component_connection.cu (hunk 3 above)
Comment (guilhermeAlmeida1) [reply device/cuda/src/cca/component_connection.cu:40]: Because I made this code match the "alt_cell" type, which uses links to a module collection to store the geometry information, not the actual geometry ids

Comment (krasznaa) [reply device/cuda/src/cca/component_connection.cu:40]: :thinking: Then let's rename the variable to something like `module_link`. As `module_id` would suggest a different content.

**Comment by guilhermeAlmeida1** [comment]
This should be ready to be looked at again.

**Review by stephenswat — APPROVED** [review]
Let's go ahead and put this in.
